spec/unit/knife/core/cookbook_scm_repo_spec.rb

```diff
@@ -59,10 +59,23 @@
 
       it "exits when there is no git repo" do
         allow(::File).to receive(:directory?).with(/.*\.git/).and_return(false)
+        allow(::File).to receive(:exists?).with(/.*\.git/).and_return(false)
         expect {@cookbook_repo.sanity_check}.to raise_error(SystemExit)
       end
+     
 
       describe "and the repo is a git repo" do
+
+        describe "and the repo has a redirected git folder" do
+          it "exits if the redirected folder does not exist" do          
+            allow(::File).to receive(:directory?).with(/.*\.git/).and_return(false)
+            allow(::File).to receive(:exists?).with(/.*\.git/).and_return(true)
+            allow(::IO).to receive(:read).with(/.*\.git/).and_return('gitdir: /Users/someone/gitrepos/thiscookbook')
+            allow(::File).to receive(:directory?).with('/Users/someone/gitrepos/thiscookbook').and_return(false)          
```

Comment:
This is hella nasty file system stubbing and ultimately this won't end well.

When we see this it means there's too much collaboration between the objects -- in this case the provider and the underlying filesystem itself.

A better approach is to setup real tempfiles and tempdirs and test the provider against those.  Expand the System Under Test to include the filesystem and the provider.

If there's files like /etc/passwd or something that are hardcoded into the provider, then those can be modified so that you can dep inject a tempfile/tempdir in place of the hard coded file.  I've used const's before in the debian ifconfig provider to do this:

https://github.com/chef/chef/blob/master/lib/chef/provider/ifconfig/debian.rb#L27-L28
https://github.com/chef/chef/blob/master/spec/unit/provider/ifconfig/debian_spec.rb#L65-L68